### DISCUSSION TITLE
How can I use Windows Terminal instead of Cmder?

### DISCUSSION THREAD

**Discussion by charlesroper** [leokhoa/laragon#369] [Q&A] [answered]
I'd like to use my own Windows Terminal instead of Cmder. Is there a way of doing this? Cmder is pretty good, but my own CLI in Windows Terminal is customised for my needs, I have my own shortcuts and prefs configured in WT, and I use Nushell, not cmd/clink. I spend quite a bit of time on the CLI and having to use Cmder isn't ideal.

When you open Cmder in Laragon, how is it loading its own "containerized" environment? Can this be recreated using a script or batch file?

What I've done in the meantime is simply add the `nu` command to the end of my `C:\tools\laragon\bin\cmder\config\user_profile.cmd` file. This loads Nushell with my custom config, but it's still using Cmder, not Windows Terminal.

Any ideas? 

Many thanks.

**Comment by charlesroper** [comment]
I've done a bit of further investigation. If I run `C:\tools\laragon\bin\cmder\vendor\init.bat` in Windows Terminal, this seems to initialise my Laragon environment. That's pretty good and I can live with that, but I'm now wondering, is there a way of uninitialising my Laragon environment? A sort of `uninit.bat`? It's not a huge problem if not - I can always just close the tab in WT and/or open a new one.

**Comment by charlesroper** [comment, marked as answer]
OK, I've added a Laragon profile in Windows Terminal and it's got me pretty close to where I want to be. Here's a screenshot:

![2023-01-11_02](https://user-images.githubusercontent.com/176/211700752-a2cd6340-c121-400c-8f49-f172da78d1dd.png)

Here's a screenshot of the profile settings in the WT UI (**Update**: you should put `cmd /k` in front of the command line, which is not shown in the screenshot):

![2023-01-11_03](https://user-images.githubusercontent.com/176/211700811-01753a89-d08c-4a57-96ed-a206089bec35.png)

Here's the same config taken directly from the WT's `settings.json` file:

```json
{
    "commandline": "cmd /k C:\\tools\\laragon\\bin\\cmder\\vendor\\init.bat",
    "guid": "{396817ac-381c-4896-a0cf-7c8ea12adf87}",
    "hidden": false,
    "icon": "C:\\Users\\Charles\\OneDrive\\Pictures\\laragon.png",
    "name": "Laragon",
    "startingDirectory": "%USERPROFILE%",
    "tabTitle": "Laragon"
}
```

Finally, here's a copy of the `laragon.png` file I used for the icon:

![laragon](https://user-images.githubusercontent.com/176/211701214-b1635bd3-0fa2-477f-9578-54e506dc7d08.png)

**Reply by charlesroper** [reply]
Here is a screenshot and quick video showing Windows Terminal in action with Laragon; my normal environment on the left, and Laragon on the right. I'm using split panes to compare the different environments side-by-side. I've added a custom theme ([Cobalt2](https://windowsterminalthemes.dev/?theme=Cobalt2)) to make it immediately clear what environment I'm in. Seems to be working well.

If anyone has any questions or needs help getting set up with something like this, just drop me a reply here.

![2023-01-13_01](https://user-images.githubusercontent.com/176/212364808-4fce7c59-e5b1-41c2-9747-471c36b20ee9.png)

https://user-images.githubusercontent.com/176/212364841-3da0f274-e3ee-4b3c-8bef-cb02654b0087.mp4

**Comment by erobertsme** [comment]
I tried this and found that the process exits immediately. Following this info here (https://github.com/cmderdev/cmder/wiki/Seamless-Windows-Terminal-Integration) I was able to get it to work by changing the commandLine setting to `cmd.exe /k \"C:\\tools\\laragon\\bin\\cmder\\vendor\\init.bat\"`

**Reply by charlesroper** [reply]
Good catch. I use [nushell](https://www.nushell.sh/) so I've put `nu.exe` as a startup command in  `%CMDER_ROOT%\config\user_profile.cmd`. This is preventing it from automatically closing for me. Putting `cmd /k` in from of Windows Terminal's `commandLine` key for the profile is a more resilient approach.

I'll update my advice above.

**Comment by rescenic** [comment]
Thank you so much for Windows Terminal setting.

**Comment by trananhmanh89** [comment]
Hi,

Here is how I implemented laragon path to my windows terminal profile

1. Create a file named "laragon.ps1" in your "laragon/usr" folder with the following content:

```powershell
$file = Join-Path $PSScriptRoot 'laragon.ini'
$laragonDir = (Get-Item $PSScriptRoot).Parent.FullName

$foundPhp = Select-String -Path $file -Pattern 'Version=(php-.+)'

if ($foundPhp) {
    $phpVersion = $foundPhp.Matches.Groups[1].Value
    $phpPath = $laragonDir + '\bin\php\' + $phpVersion
    $env:Path = $phpPath + ';' + $env:Path
}

$foundNode = Select-String -Path $file -Pattern 'Version=(node-.+)'

if ($foundNode) {
    $nodeVersion = $foundNode.Matches.Groups[1].Value
    $nodePath = $laragonDir + '\bin\nodejs\' + $nodeVersion
    $env:Path = $nodePath + ';' + $env:Path
}
```

2. Create a new profile in Windows Terminal.

3. Set the following command as the "Command line" option: `powershell -NoExit -ExecutionPolicy Bypass -File E:/laragon/usr/laragon.ps1` (Replace "E:/laragon/usr/laragon.ps1" with the correct path to your "laragon.ps1" file on your PC).

4. Save the profile.

5. Create new session with that profile

6. Test the PHP and Node.js versions in the Windows Terminal.

Hope this help.
